drivers/net/ethernet/renesas/rswitch.c

```diff
@@ -4299,6 +4299,54 @@ static void rswitch_fwd_init(struct rswitch_private *priv)
 		   priv->addr + FWCMPTC);
 }
 
+static void rswitch_set_max_hash_collisions(struct rswitch_private *priv)
+{
+	/* 320Mhz */
+	const u64 clk_freq = 320 * 1000 * 1000;
+	/* Avarage frame size in bits */
+	const u64 average_frame_size = 64 * 8;
+	u64 tsn_throughput = 0, max_throughput;
+	struct device_node *ports, *port, *phy = NULL;
+	int err = 0;
+
+	ports = of_get_child_by_name(priv->pdev->dev.of_node, "ports");
+	if (!ports) {
+		/* Set minimum value for collision number */
+		priv->max_collisions = 1;
+		return;
+	}
+
+	for_each_child_of_node(ports, port) {
+		phy = of_parse_phandle(port, "phy-handle", 0);
+		if (phy) {
+			/* 1 GBit*/
+			tsn_throughput += 1000 * 1000 * 1000;
+		} else {
+			if (of_phy_is_fixed_link(port)) {
+				struct device_node *fixed_link;
+				u32 link_speed;
+
+				fixed_link = of_get_child_by_name(port, "fixed-link");
+				err = of_property_read_u32(fixed_link, "speed", &link_speed);
+				if (err)
+					continue;
+
+				tsn_throughput += link_speed * 1000 * 1000;
+			}
+		}
+	}
+
+	max_throughput = tsn_throughput + priv->gwca.speed * 1000 * 1000;
+
+	/* Calculate maximum collisions number using the formula:
+	 * FWLTHHEC.LTHHMC =
+	 * (clk_freq[Hz] * Average_frame_size[bit] / Incoming_throughput[bps] - 4) / 3
+	 */
+	priv->max_collisions = (((clk_freq * average_frame_size) / (max_throughput)) - 4) / 3;
+
+	of_node_put(ports);
```

Comment:
It is better to release `ports` right after use.  I.e. near line 4338

Reply:
Good point, moved.